### PR TITLE
Implementation of deletion of Posts and Events

### PR DESCRIPTION
**Pull request by Antoine444**
This PR introduces the implementation of the deletion of `Event`s and `Post`s. This is a significant PR for our app since now the `creator` of an event can delete posts he previously wrote and more importantly he can delete an `Event` he created in a safely manner (meaning all fields in `Firebase` are updated correctly when the `Event` is deleted. The fields that might need an update when an `Event` is deleted are the `myEvents`, `joinedEvents`, `myFavorites` and `pendingRequests` of a `GoMeetUser`).

To do so, we added `delete` icons for an `Event` and a `Post` and when the user clicks on it, it displays a `AlertDialog` that asks the user if he is sure about the deletion. When the deletion happened, the UI is updated as well.

Screenshot 2024-05-26 at 6 01 33 PM

### CODE REVIEW

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt [hunk 1]
```diff
@@ -244,15 +260,25 @@ fun MyEventInfo(
                   Spacer(Modifier.height(screenHeight / 50))
                 } else {
                   Spacer(Modifier.height(10.dp))
-                  myEvent.value!!.posts.forEach { post ->
+                  posts.reversed().forEach { post ->
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt:263]: I don't understand why each time posts is updated there is a call to reversed() when there is also a call to reversed() upon displaying the list of posts. Logically if reversed() is called upon initialisation/modification of posts, but also upon the usage of posts, wouldn't it make more sense to remove each reversed() call? (I mean that since the list gets reversed twice then it ends up as in the same order as if it weren't reversed at all, so what's the point of all the reverse calls?)

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt [hunk 2]
```diff
@@ -266,6 +292,37 @@ fun MyEventInfo(
   if (showShareEventDialog) {
     ShareDialog("Event", eventId, onDismiss = { showShareEventDialog = false })
   }
+
+  if (showDeleteEventDialog) {
+    DeleteEventDialog(
+        onConfirm = {
+          coroutineScope.launch {
+            myEvent.value!!.participants.forEach { participant ->
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt:300]: What happens in the onConfirm lambda should really be a function in EventViewModel, but technically that's a convention more than anything else, so as long as it works it's all good and we can refactor it later.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt [hunk 3]
```diff
@@ -266,6 +292,37 @@ fun MyEventInfo(
...
+            }
+
+            myEvent.value!!.pendingParticipants.forEach { pendingParticipant ->
+              userViewModel.invitationCanceled(eventId, pendingParticipant)
+            }
+
+            val allUsers = userViewModel.getAllUsers()
+
+            allUsers!!.forEach { user ->
+              if (user.myFavorites.contains(eventId)) {
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt:315]: Shouldn't each users' `joinedEvents` list also get checked and potentially updated here upon the deletion of the events?

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt [hunk 4]
```diff
@@ -266,6 +292,37 @@ fun MyEventInfo(
...
+                userViewModel.removeFavoriteEvent(eventId, user.uid)
+              }
+            }
+
+            eventViewModel.removeEvent(eventId)
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt:320]: This viewmodel function should incorporate the whole logic of this lambda

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt (hunk 3 above)
Comment (DanThePol) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt:315]: My bad, I realised that `gotKickedFromEvent` handles that.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt [hunk 5]
```diff
@@ -63,14 +69,17 @@ fun EventPost(
     post: Post,
     userViewModel: UserViewModel,
     eventViewModel: EventViewModel,
-    currentUser: String
+    currentUser: String,
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:72]: currentUser shouldn't be passed as a parameter anymore if there is an EventViewModel or a UserViewModel passed as well. Both viewmodels have access to the current user via ...ViewModel.currentUID!! which is guaranteed to be non-null as long as it is the instance passed from InitFunctions.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt [hunk 6]
```diff
@@ -173,12 +182,67 @@ fun EventPost(
...
+                imageVector = Icons.Default.Delete,
+                contentDescription = "Delete",
+                tint = MaterialTheme.colorScheme.onBackground)
+          }
+        }
+
         Spacer(modifier = Modifier.weight(1f))
         Text(
-            text = "${event.getDateString()}, ${event.getTimeString()}",
+            text = "${event.date}, ${event.time}",
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:196]: Why the change from using .getDateString() and .getTimeString() to this? Those functions ensure that the date and time as Strings are consistent everywhere in terms of format ("DD/MM/YYYY" or "Today" for date and "HH:MM" in 24h for time), whereas using the implicit toString function of LocalDate and LocalTime might not necessarily yield the same.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt [hunk 7]
```diff
@@ -173,12 +182,67 @@ fun EventPost(
...
             style = MaterialTheme.typography.bodyMedium,
             color = Grey)
       }
     }
   }
+
+  if (showDeletePostDialog) {
+    DeletePostDialog(
+        onConfirm = {
+          coroutineScope.launch {
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:206]: Now this is what I like to see. Clean, concise and every function is in the right class (according to me in any case)!

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt (hunk 6 above)
Comment (DanThePol) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:196]: Screenshot 2024-05-26 at 23 03 14
Now the date/time looks like this (bottom right) rather than DD/MM/YYYY HH:MM

File: app/src/main/java/com/github/se/gomeet/viewmodel/EventViewModel.kt [hunk 8]
```diff
@@ -428,6 +428,22 @@ class EventViewModel(val currentUID: String? = null) : ViewModel() {
         event.copy(pendingParticipants = event.pendingParticipants.minus(userId)))
   }
 
+  /**
+   * Update the posts field of the event by removing the given post to the list.
+   *
+   * @param event the event to update
+   * @param post the post to remove from the event
+   */
+  fun deletePost(event: Event, post: Post) {
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/viewmodel/EventViewModel.kt:437]: Ideally this function would update only the `posts` list in the concerned event rather than the whole event (unnecessary network traffic), but again, as long as it works for now.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt (hunk 6 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:196]: Because ChatGPT trolled me when I asked him to clean the code a little bit

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt [hunk 9]
```diff
@@ -196,6 +196,37 @@ class UserViewModel(val currentUID: String? = null) : ViewModel() {
...
   }
 
+  /**
+   * User deletes an event and removes it from their list of myEvents. It is used when a user
+   * deletes an event.
+   *
+   * @param eventId The id of the event to delete.
+   * @param userId The id of the user deleting the event.
+   */
+  suspend fun userDeletesEvent(eventId: String, userId: String = currentUID!!) {
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:206]: This is something that we have done from the start but is a mistake. Suspend functions shouldn't exist in ViewModels, because they get called in Composable functions. I'm not sure whether it really changes anything performance-wise, but I discovered that if you want to call suspend functions in a class extending the basic ViewModel class, you can use `viewModelScope.launch { /* async stuff going on */ }`. This would be a lot cleaner and more correct convention-wise. From my understanding, it is also the case that a suspend function launched in a coroutine will get cancelled as soon as that coroutine "dies" or is refreshed. I.e. when you call a suspend function in a coroutineScope inside a Composable function, then as soon as that Composable gets refreshed (or if the user navigates to another screen), the suspend function launched in the coroutineScope is cancelled if it hasn't completed yet.

File: app/src/main/java/com/github/se/gomeet/viewmodel/EventViewModel.kt (hunk 8 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/viewmodel/EventViewModel.kt:437]: I don't understand what you are saying. To update the list of posts I need to have access to the `Event` where this list is stored.

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt [hunk 10]
```diff
@@ -196,6 +196,37 @@ class UserViewModel(val currentUID: String? = null) : ViewModel() {
...
+    }
+  }
+
+  /**
+   * Removes the given event from the user's list of favorites.
+   *
+   * @param eventId The id of the event to remove from favorites.
+   * @param userId The id of the user for which we remove the event from favorites.
+   */
+  suspend fun removeFavoriteEvent(eventId: String, userId: String = currentUID!!) {
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:221]: Again, for both of these functions it would be better (in my opinion) to just update the corresponding list rather than the whole user on Firebase, but this works too.

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt (hunk 9 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:206]: Well, it works for now we will do the refactoring later

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt (hunk 10 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:221]: Again, I don't understand what you are saying. To update the `myFavorites`field of a `GoMeetUser`, I need to update the `GoMeetUser`. If I don't then in Firebase the user would have in his `myFavorites` field, an `Event` that doesn't exist anymore if this `Event` is deleted.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt (hunk 5 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:72]: I just used the `EventPost` function that was implemented in a previous PR by another developper.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt (hunk 4 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt:320]: Yes but to have the whole logic in here, the `EventViewModel` would have needed a `UserViewModel` and I wasn't sure if it was good practise to call another viewModel inside of a viewModel when implementing the needed functions in `EventViewModel`.

**Review by DanThePol — CHANGES_REQUESTED** [review]
Functionality-wise everything is ok, but I think that the post date and time format should be reverted to the what it originally was (using event.getDateString and event.getTimeString). The rest of the comments are more details that can be ironed out later.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt (hunk 4 above)
Comment (DanThePol) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt:320]: I suppose that makes sense. I don't know if what I'm about to suggest is good practice or not, but what I would've done is put the function in EventViewModel and then made it communicate directly with another static UserRepository function, rather than going through the UserViewModel from the EventViewModel. But then again, it might be bad practice to access the UserRepository from the EventViewModel. Object-oriented is confusing sometimes ngl

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt (hunk 5 above)
Comment (DanThePol) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:72]: My bad, it showed up as a modification in the git differences and at this particular difference a bit quickly and so I thought you had added it yourself 😅

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt (hunk 6 above)
Comment (DanThePol) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:196]: Ahhh sacré GPT 🤖

File: app/src/main/java/com/github/se/gomeet/viewmodel/EventViewModel.kt (hunk 8 above)
Comment (DanThePol) [reply app/src/main/java/com/github/se/gomeet/viewmodel/EventViewModel.kt:437]: It's possible to update only one field of a document in firebase, rather than updating the whole document. And in this situation, since you are modifying only the `posts` field of a document in the `events` Firebase folder, it would make more sense to reupload only the `posts` field rather than the whole event.

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt (hunk 10 above)
Comment (DanThePol) [reply app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:221]: See my previous comment response about this

Comment (DanThePol) [reply app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:221]: In UserRepository, we update the entire user like so:
`// ...
val documentRef = Firebase.firestore.collection(USERS_COLLECTION).document(user.uid)
      documentRef
          .update(user.toMap())`

But for instance, if you have a look at the updateRating function in EventRepository, only the ratings field is updated:

`// ...
val documentRef = Firebase.firestore.collection(EVENT_COLLECTION).document(eventID)
      documentRef
          .update(RATINGS, ratings)`

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt (hunk 1 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/MyEventInfo.kt:263]: I do need all of them, I tried to retrieve them cases by cases and I need them, otherwise when creating or deleting new posts, it won't keep them in reversed order. I do need the 4 reversed() call I made, you can try it by yourself if you want to convince yourself.

File: app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt (hunk 6 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/ui/mainscreens/events/posts/EventPost.kt:196]: Should be fixed now

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt (hunk 10 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:221]: Ok my bad, I didn't see the implementation of `updateRating` function, I took inspiration from other functions in the viewModel that were implemented in the same logic that I did in this PR.

File: app/src/main/java/com/github/se/gomeet/viewmodel/EventViewModel.kt (hunk 8 above)
Comment (Antoine444) [reply app/src/main/java/com/github/se/gomeet/viewmodel/EventViewModel.kt:437]: Got it, will do it for next time, thanks.

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt [hunk 11]
```diff
@@ -45,7 +45,8 @@ class UserViewModel(val currentUID: String? = null) : ViewModel() {
       email: String,
       phoneNumber: String,
       country: String,
-      pfp: String = ""
+      pfp: String = "",
+      favorites: List<String> = emptyList()
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:49]: It doesn't make sense to introduce this here, since upon the creation of a user that user won't have any events. This technically doesn't hurt though and it doesn't change the code functionality, so I'll approve the PR.

File: app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt [hunk 12]
```diff
@@ -63,9 +64,9 @@ class UserViewModel(val currentUID: String? = null) : ViewModel() {
                   email = email,
                   phoneNumber = phoneNumber,
                   country = country,
-                  joinedEvents = emptyList(),
+                  joinedEvents = favorites,
```
Comment (DanThePol) [inline app/src/main/java/com/github/se/gomeet/viewmodel/UserViewModel.kt:67]: Also it doesn't make sense that joinedEvents = favorites

**Review by DanThePol — APPROVED** [review]
Issues resolved, approved.
